### PR TITLE
[ILLink analyzer] Move generic parameter analysis to dataflow

### LINKED ISSUE #95121: Trim analyzer: Doesn't produce warnings due to annotated generic parameter in various situations
```csharp
Method<string>();

Type Method<TUnknown>()
{
    // Should WARN - TUnknown does have All annotation
    return typeof(TRequires<TUnknown>);
}

class TRequires<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] T>
{
}
```

The able produces IL2091 in the trimmer, but doesn't produce it in the analyzer. The bug seems to be that we don't process generic parameter annotations in a `typeof`.

Also happens for a declaration of a local variable:
```csharp
void Method<TUnknown>()
{
    // Trimmer produces IL2091 here, analyzer doesn't
    TRequires<TUnknown> a;
}
```

Note that for local variables NativeAOT doesn't produce the warning either - see the comment here for why:
https://github.com/dotnet/runtime/blob/7f818485a669eef2271034cb1aee3b01bdeeaef6/src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/GenericParameterWarningLocation.cs#L12-L28

We should figure out if analyzer needs to produce the warning or not.

Another situation is a static field access:
```csharp
    TRequires<TUnknown>.Field;  // Should warn
```

Yet another situation - expression trees:
```csharp
Expression<Func<string>> a = () => TRequires<TUnknown>.Field
```

And then type checks (NativeAOT doesn't warn here, see the link above):
```csharp
if (value is TRequires<TUnknown>) {}

var a = value as TRequires<TUnknown>;
```

Catch and filter expressions (NativeAOT doesn't warn):
```csharp
try {}
catch (TRequires<TUnknown>) {}
catch (Exception ex) when (ex is TRequires<TUnknown>) {}
```

Nested generic for methods:
```csharp
GenericMethod<Generic1<Generic2<TRequires<TUnknown>>>();
```

Similarly nested generics don't cause DAM marking and thus diagnostics:
```csharp
GenericMethod<Generic1<Generic2<TRequires<TypeWIthRUCMethod>>>(); // Should produce IL2026
```

### PR DESCRIPTION
**Pull request by sbomer**
And align behavior more closely with NativeAot. I tried to roughly follow the implementation approach used by NativeAot in GenericArgumentDataFlow.cs to make it easier to share the logic (without actually sharing the code for now).

Fixes https://github.com/dotnet/runtime/issues/95121.

Some things to note:
- Gets rid of some analyzer warnings in cases where NativeAot doesn't warn either
- Doesn't warn for generics in `typeof`
- Doesn't warn for generics in signatures of reflectable members (ILC does this to work around an incorrect suppression in DI, as discussed in https://github.com/dotnet/runtime/issues/81358).
- Generics in base types or interface types are still analyzed outside of dataflow

### CODE REVIEW

File: src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/GenericParameterDataFlow.cs [hunk 1]
```diff
@@ -406,7 +406,7 @@ class TypeGenericRequirementsOnMembers<[DynamicallyAccessedMembers (DynamicallyA
 		{
 			public TypeRequiresPublicFields<TOuter> PublicFieldsField;
 
-			[ExpectedWarning ("IL2091", nameof (TypeRequiresPublicMethods<TOuter>), ProducedBy = Tool.Trimmer | Tool.Analyzer)] // NativeAOT_StorageSpaceType
+			[ExpectedWarning ("IL2091", nameof (TypeRequiresPublicMethods<TOuter>), ProducedBy = Tool.Trimmer)] // NativeAOT_StorageSpaceType
```
Comment (vitek-karas) [inline src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/GenericParameterDataFlow.cs:409]: Nit: Maybe update the comment about NativeAOT_StorageSpaceType thing at the top of the file to also mention that we now expect the same behavior from the analyzer.

**Review by vitek-karas — APPROVED** [review]
Nice!

**Comment by ShreyasJejurkar** [comment]
This is great. I just had one question can't we share the trimming logic between the trimmer, NativeAOT, and the rest of the app models where we do trimming? This is hard to do or is it technically impossible to do so?

**Comment by vitek-karas** [comment]
The additional complication is that trimmer (illink) and NativeAOT see IL, while the analyzer sees the Roslyn's operations/CFG. We could probably share more if(when) we implement CFG reconstruction from IL in trimmer and NativeAOT. But currently the two approaches have to do some things differently.

That said, we already share quite a lot of the core data flow logic, and almost all of the tests.

**Pull request merged**